Test Merge Reports. Closes #196

abaiken opened the pull request:
I attempted to write these new tests to only depend on the SCAN_DATA collected when you run scans. In doing so I built the tests around scans that already existed/or that I added to my config file.

The test `test_merge_reports_from_scanjob` depends on the following scans in the config file:
```
- name: non-rhel
          sources:
              - 'sonar-none-fedora-26-vc'
              - 'sonar-none-centos-7-vc'
- name: rhel-7
          sources:
               - 'sonar-none-rhel-7-vc'
               - 'sonar-jbossas-rhel-7-vc'
               - 'sonar-jbossamq-rhel-7-vc'`
```
If they are not present, it currently just returns--which results in a passing test. (Open to suggestions to either have this skip instead --> @pytest.mark.skipif(reason=SCAN_DATA == {}), or to create the scans/scanjobs inside the function)

The test `test_merge_reports_negative` would ideally have the following scans in the config file:
```
- name: Unreachable
         sources:
             - 'unreachable'
- name: Connection
         sources:
             - 'sonar-jbosseap-rhel-5-vc'
         type: 'connect'
```
If they are not present, it simply does not add them to the list of invalid identifiers to check for.

abaiken committed:
create Report object, add tests for merge reports

abaiken committed:
fix docstring

abaiken committed:
Merge branch 'master' of https://github.com/quipucords/camayoc into issues/196

abaiken committed:
update file paths

abaiken committed:
fix import error

kdelee commented:
@abaiken I think it would be better if they skipped. You can do this from within the test using some functionality in py.test. Here is a link to the documentation.

https://docs.pytest.org/en/documentation-restructure/how-to/skipping.html

I think that what we need to do with these tests that depend on the scans being run is:
* Skip if the scans were just not run at all
* "xfail" if the scan was run but did not complete some stage necessary for the test (for example, the scan was run but for some reason we don't have a report and we need the report in order to run specific test. Then the test should xfail).

If we can do that to all existing and future tests that depend on `SCAN_DATA`, then I think our results will make alot more sense. Will also close #214 , #211, #212

kdelee commented on camayoc/qpc_models.py:
```diff
@@ -592,3 +593,75 @@ def equivalent(self, other):
+
+class Report(QPCObject):
+    """A class to aid in the creation and control of Reports."""
+
```
Can you add into this doc string how you can get the report id from a scan?

Also example of how you might create a report object and use to merge results from multiple scans.

kdelee reviewed (commented):
Left an initial comment.

I'm on the fence about the `assert_merge_fails` being an instance method. I guess my only substantive objection is that may not be used by any other tests, while the `Report` object certainly seems like it could get used in other tests (thanks :1st_place_medal: ). :man_shrugging:

abaiken committed:
Merge branch 'master' of https://github.com/quipucords/camayoc into issues/196

abaiken committed:
move assert merge fails out of class, update Report docstring, skip tests if RUN_SCANS = False

abaiken committed:
Fix docstring

abaiken committed:
fix import order

abaiken commented:
@kdelee The tests now skip if `RUN_SCANS` is set to `False`.  The `test_merge_reports_from_scanjob` xfails if the two scans it depends on are not in the config file. I moved `assert_merge_fails` out of the `Report` class and into the `test_reports` file. Also, I updated the docstring for the Report object and provided an example of how to create a Report object from merging.

abaiken committed:
Merge branch 'master' of https://github.com/quipucords/camayoc into issues/196

kdelee reviewed (approved):
@abaiken LGTM

Thanks for those updates. Chris just merged a PR that turns the `skipif` thing into a fancy decorator you can import from `camayoc/tests/qpc/utils` called "mark_runs_scans"

You import it and apply it on top of the test declaration with `@mark_runs_scans`.

Could be done in follow up PR or this. Think it could get applied to some CLI tests as well so not a big deal  if you want to leave it till later

abaiken committed:
closes #196

abaiken commented:
@kdelee Done.

chambridge commented:
🍾 🎉 🎈
